text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,74 @@
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a constant evaluation or runtime constness-context. Language features and standard library functions may also differ in behavior depending on the constness-context they have been used or called in.
+
+The constness-context of a function is therefore observable behavior.
+
+If a `const fn` is called in a runtime constness-context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks to this RFC.
+
+Firstly, pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime constness-context, as long as it behaves exactly like runtime.
+
+Secondly, with the current rules around `const fn` purity, unsafe code could choose to only sometimes call unknown `const fn` code at runtime depending on whether it needs the result. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time".
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternative is to keep the current rules. This is bad because the current rules are highly restrictive and don't have a significant benefit to them.
+
+It would also be possible to allow them to behave differently, but keep the restrictions around purity and determinism at runtime. This would still allow unsafe code to treat `const fn` specially, but this is not seen as a desirable feature of `const fn`.
+
+# Prior art
+[prior-art]: #prior-art
+
+C++ with `constexpr`, a compile-time evaluation system similar to Rusts `const fn`, has a [`std::is_constant_evaluated`](std-is-constant-evaluated) function which can be used to determine whether the function is being executed during constant evaluation or at runtime. It does not impose restrictions that code has to behave the same during constant evaluation or runtime.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Do we want to allow `const fn` evaluated during constant evaluation to behave nondeterministically?
```

Comment:
That's already impossible. The const evaluator is completely deterministic and that is required for type system soundness. (`let x: [u8; C] = [42; C];` must always evaluate both `C` to the same value)